### PR TITLE
Improve the documentation of SSL support for TDS connections

### LINKED ISSUE #99: Improve the documentation of SSL support for TDS connections
Currently, there is no documentation for $subject. Here are my thoughts for the initial draft.

1. With SSL support compiled in, the PostgreSQL server needs to be started with SSL enabled by setting the parameter ssl to on in postgresql.conf. Otherwise, SSL support can't be enabled. In case, any client requests for SSL connection, server will reject the connection.
2. TDS extension uses the same OpenSSL configrations as psql connections.
3. There are different options to initiate a SSL connection from a SQL Server client driver. Based on the option, Babelfish server behaves differently. This link documents the current behavior - https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/babelfish-configuration.html#babelfish-ssl.

### PR DESCRIPTION
**Pull request by kuntalghosh**
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

### Description
This commit improves the documentation for end-to-end encryption of TDS connections.
 
### Issues Resolved
#99

### Check List
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _internals/tds-protocol.md [hunk 1]
```diff
@@ -85,8 +85,27 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
 <code>LOGIN7</code> which defines the authentication rules for use 
 between client and server.
 
-
-### Defining authentication rules.
+### SSL/TLS support for TDS connections
+
+Babelfish uses native PostgreSQL support for using SSL connections to encrypt client/server communications. See PostgresSQL [documentation](https://www.postgresql.org/docs/13/ssl-tcp.html#SSL-SETUP) for details about configuring server-side SSL functionality.
```
Comment (susanmdouglas-aws) [inline _internals/tds-protocol.md:90]: I see that this has been pushed, but there is a typo in the above line... PostgreSQL is misspelled.

Comment (kuntalghosh) [reply _internals/tds-protocol.md:90]: No Susan, the commit wasn't pushed in the main branch. I've only pushed it in a separate branch. Once I fix all the review comments and get approvals, I'll merge the changes.

Comment (susanmdouglas-aws) [reply _internals/tds-protocol.md:90]: Sorry, my bad!  Let me try again:

Babelfish uses native PostgreSQL functionality to support SSL connections that encrypt client/server communications. See the PostgreSQL [documentation](https://www.postgresql.org/docs/13/ssl-tcp.html#SSL-SETUP) for details about configuring server-side SSL functionality.

File: _internals/tds-protocol.md [hunk 2]
```diff
@@ -85,8 +85,27 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
...
+With SSL support compiled in, the PostgreSQL server can be started with SSL enabled by setting the parameter <code>ssl</code> to <code>on</code> in postgresql.conf. The server will listen for both normal and SSL connections on the same TCP port, and will negotiate with any connecting client on whether to use SSL. By default, a client can decide whether to use SSL connections or not. If <code>babelfishpg_tds.tds_ssl_encrypt</code> is set to <code>true</code>, end-to-end encryption is enforced for all connections. In that case, if a client requests for a non-encrypted connection, the connection is rejected.
```
Comment (susanmdouglas-aws) [inline _internals/tds-protocol.md:91]: How about:

To use SSL encryption,  [compile the PostgreSQL server with support for SSL](https://www.postgresql.org/docs/13/install-procedure), and set the <code>ssl</code> parameter to <code>on</code> in the postgresql.conf file. The server will listen for SSL and non-SSL connections on the same TCP port, and will negotiate with any connecting client about using SSL. By default, a client can decide whether to use SSL connections or not. If <code>babelfishpg_tds.tds_ssl_encrypt</code> is set to <code>true</code>, end-to-end encryption is enforced for all connections. In that case, if a client requests a non-encrypted connection, the connection is rejected.

File: _internals/tds-protocol.md [hunk 3]
```diff
@@ -85,8 +85,27 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
...
+
+The table following shows how Babelfish behaves for each combination.
```
Comment (susanmdouglas-aws) [inline _internals/tds-protocol.md:93]: How about:

The following table shows how Babelfish interprets SSL settings when a client connects:

File: _internals/tds-protocol.md [hunk 4]
```diff
@@ -85,8 +85,27 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
...
+| ENCRYPT_NOT_SUP      | tds_ssl_encrypt=true  | No, connection closed                       | ENCRYPT_REQ              |
+| ENCRYPT_REQ          | tds_ssl_encrypt=false | Allowed, the entire connection is encrypted | ENCRYPT_ON               |
+| ENCRYPT_REQ          | tds_ssl_encrypt=true  | Allowed, the entire connection is encrypted | ENCRYPT_ON               |
+| ENCRYPT_CLIENT_CERT  | tds_ssl_encrypt=false | No, connection closed                       | Unsupported              |
+| ENCRYPT_CLIENT_CERT  | tds_ssl_encrypt=true  | No, connection closed                       | Unsupported              |
+
+### Defining authentication rules
 
 <code>LOGIN7</code> is the TDS ways of defining authentication rules. 
 Client and server both have to know how to handle authentication rules, which are vital to security.
```
Comment (susanmdouglas-aws) [inline _internals/tds-protocol.md:111]: Missing article:

The client and server both...

**Comment by 3manuek** [comment]
Updated this branch with the PostgreSQL misspelling fix. cc / @kuntalghosh @gopinathpai

File: _internals/tds-protocol.md (hunk 1 above)
Comment (3manuek) [reply _internals/tds-protocol.md]: [Fixed](https://github.com/babelfish-for-postgresql/babelfish_project_website/pull/145/commits/d6ef975eeca30403245ee37cb8299e18e54c51e4)

File: _internals/tds-protocol.md (hunk 2 above)
Comment (kuntalghosh) [reply _internals/tds-protocol.md]: [Fixed](https://github.com/babelfish-for-postgresql/babelfish_project_website/pull/145/commits/19d2cec5197928d7eb6ec4ae291f6a82d150861f)

File: _internals/tds-protocol.md (hunk 4 above)
Comment (kuntalghosh) [reply _internals/tds-protocol.md]: [Fixed](https://github.com/babelfish-for-postgresql/babelfish_project_website/pull/145/commits/19d2cec5197928d7eb6ec4ae291f6a82d150861f)

File: _internals/tds-protocol.md [hunk 5]
```diff
@@ -85,11 +85,31 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
 <code>LOGIN7</code> which defines the authentication rules for use 
 between client and server.
 
+### SSL/TLS support for TDS connections
 
-### Defining authentication rules.
+Babelfish uses native PostgreSQL support for using SSL connections to encrypt client/server communications. See PostgreSQL [documentation](https://www.postgresql.org/docs/13/ssl-tcp.html#SSL-SETUP) for details about configuring server-side SSL functionality.
```
Comment (susanmdouglas-aws) [inline _internals/tds-protocol.md:90]: The start of the second sentence is missing the word 'the' - it should be 'See the PostgreSQL documentation...'

File: _internals/tds-protocol.md [hunk 6]
```diff
@@ -85,11 +85,31 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
...
+
+To use SSL encryption, [compile the PostgreSQL server with support for SSL,](https://www.postgresql.org/docs/13/install-procedure) and set the <code>ssl</code> parameter to <code>on</code> in the postgresql.conf file. The server will listen for SSL and non-SSL connections on the same TCP port, and will negotiate with any connecting client about using SSL. By default, a client can choose whether to require an SSL connection or not. The server will listen for both unencrypted and encrypted connections on the same TCP port, and negotiate the connection type with the connecting client.  If <code>babelfishpg_tds.tds_ssl_encrypt</code> is set to <code>true</code>, end-to-end encryption is enforced for all connections. In that case, if a client requests an unencrypted connection, the connection is rejected.
```
Comment (nasbyj) [inline _internals/tds-protocol.md:92]: Since we have our own install instructions we can't just point at community instructions. We need to give explicit direction for how to do this when installing (or perhaps we should just ensure that we always compile with SSL enabled).

File: _internals/tds-protocol.md [hunk 7]
```diff
@@ -85,11 +85,31 @@ Once we are over the <code>PRELOGIN</code> phase, we are ready for
...
+| ENCRYPT_NOT_SUP      | tds_ssl_encrypt=false | Yes                                         | ENCRYPT_NOT_SUP          |
+| ENCRYPT_NOT_SUP      | tds_ssl_encrypt=true  | No, connection closed                       | ENCRYPT_REQ              |
+| ENCRYPT_REQ          | tds_ssl_encrypt=false | Allowed, the entire connection is encrypted | ENCRYPT_ON               |
+| ENCRYPT_REQ          | tds_ssl_encrypt=true  | Allowed, the entire connection is encrypted | ENCRYPT_ON               |
+| ENCRYPT_CLIENT_CERT  | tds_ssl_encrypt=false | No, connection closed                       | Unsupported              |
+| ENCRYPT_CLIENT_CERT  | tds_ssl_encrypt=true  | No, connection closed                       | Unsupported              |
+
+### Defining authentication rules
 
 <code>LOGIN7</code> is the TDS ways of defining authentication rules.
```
Comment (nasbyj) [inline _internals/tds-protocol.md:111]: /the TDS way of defining/is how TDS defines/

However, I don't understand what we're trying to say here. What `LOGIN7` is isn't terribly clear. More importantly, why does anyone even care? AFAICT `LOGIN7` is some super low-level detail of the TDS protocol, which doesn't seem like something users should need to care about. What am I missing here? When would a user need to understand `LOGIN7`?

**Review by nasbyj — CHANGES_REQUESTED** [review]

File: _internals/tds-protocol.md (hunk 7 above)
Comment (kuntalghosh) [reply _internals/tds-protocol.md:111]: Yeah, I agree with you. We can remove this part. Let's address this in a separate pull request.

**Review by 3manuek — APPROVED** [review]

**Review by nasbyj — APPROVED** [review]
